src/bigartm/cooccurrence_dictionary.cc

```diff
@@ -0,0 +1,442 @@
+  // resulting file with all co-occurrences
+
+  // If no co-occurrence found or it's low than min_cooc_value file isn't
+  // created
+  try {
+    path_to_vw = vw;
+    if (!path_to_vw.size()) {
+      std::cerr << "input file in VowpalWabbit format not specified\n";
+      throw 1;
+    }
+    path_to_vocab = vocab;
+    if (!path_to_vocab.size()) {
+      std::cerr << "input file in UCI vocab format not specified\n";
+      throw 1;
+    }
+    window_width = wind_width;
+    cooc_min_tf = min_tf;
+    cooc_min_df = min_df;
+    cooc_tf_file_path = cooc_tf_file;
+    cooc_df_file_path = cooc_df_file;
+    calculate_tf_cooc = cooc_tf_file_path.size() ? true : false;
+    calculate_df_cooc = cooc_df_file_path.size() ? true : false;
+    FetchVocab();
+    if (dictionary.size() > 1) {
+      ReadVowpalWabbit();
+      if (batch_manager.GetBatchQuan())
+        ReadAndMergeBatches();
+    }
+  } catch (...) {}
+}
+
+inline void CooccurrenceDictionary::FetchVocab() {
+  // This func reads words from vocab, sets them unique id and collects pair
+  // in dictionary
+  std::filebuf fb;
+  if (!fb.open(path_to_vocab, std::ios::in)) {
+    std::cerr << "Failed to open vocab\n";
+    throw 1;
+  }
+  std::istream vocab(&fb);
+  int last_token_id = 1;
+  std::string str;
+
+  while (true) {
+    getline(vocab, str);
+    if (vocab.eof())
+      break;
+    boost::algorithm::trim(str);
+    if (!str.empty())
+      dictionary.insert(std::make_pair(str, last_token_id++));
+  }
+}
+
+inline void CooccurrenceDictionary::UploadBatchOnDisk(
+        BatchManager &batch_manager,
+        std::map<int, std::map<int, cooccurrence_info>> &cooc) {
+  CooccurrenceBatch *batch = batch_manager.CreateNewBatch();
+  for (auto iter = cooc.begin(); iter != cooc.end(); ++iter) {
+    batch->FormNewCell(iter);
+    batch->WriteCell();
+  }
+  delete batch;
+}
+
+inline cooccurrence_info CooccurrenceDictionary::FormInitialCoocInfo(int doc_id) {
+  cooccurrence_info res;
+  res.doc_quan = res.cooc_value = 1;
+  res.prev_doc_id = doc_id;
+  return res;
+}
+
+inline void CooccurrenceDictionary::AddInCoocMap(int first_token_id,
+        int second_token_id, int doc_id,
+        std::map<int, std::map<int, cooccurrence_info>>& cooc_map) {
+  cooccurrence_info tmp = FormInitialCoocInfo(doc_id);
+  std::map<int, cooccurrence_info> tmp_map;
+  tmp_map.insert( std::pair<int, cooccurrence_info> (second_token_id, tmp));
+  cooc_map.insert(std::pair<int, std::map<int, cooccurrence_info>> (first_token_id, tmp_map));
+}
+
+inline void CooccurrenceDictionary::ModifyCoocMapNode(int second_token_id,
+        int doc_id, std::map<int, cooccurrence_info>& map_node) {
+  auto iter = map_node.find(second_token_id);
+  if (iter == map_node.end()) {
+    cooccurrence_info tmp = FormInitialCoocInfo(doc_id);
+    map_node.insert(std::pair<int, cooccurrence_info> (second_token_id, tmp));
+  } else {
+    iter->second.cooc_value++;
```

Comment:
http://stackoverflow.com/questions/24853/what-is-the-difference-between-i-and-i

It can give problems if object++ is big (because of coping), although I think, that modern compilers can avoid it. But still ++object considered to be better style.